Bare bones widget inspector support.

jacob314 opened the pull request:
Sent out for review without tests in expectation of significant changes based on code review feedback.

Toggle the widget inspector from the flutter tool by pressing 'i'.
When the widget inspector is select mode:
Pointer down to to inspect a widget.
Pointer click to finalize selection of a widget. You can now interact
with the application as you normally would but with the inspected widget
highlighted.
Click the inspect icon in bottom left corner of screen to reactivate
select mode.

eseidelGoogle commented:
That error from the CLA bot typically means the email address you set up for git (and got used in the git commit) is not associated with your github account.

eseidelGoogle commented:
looks like you just want to set `git config --global user.email`

jacob314 committed:
Bare bones widget inspector support.

Toggle the widget inspector from the flutter tool by pressing 'i'.
When the widget inspector is select mode:
Pointer down to to inspect a widget.
Pointer click to finalize selection of a widget. You can now interact
with the application as you normally would but with the inspected widget
highlighted.
Click the inspect icon in bottom left corner of screen to reactivate
select mode.

jacob314 commented on packages/flutter/lib/src/widgets/app.dart:
```diff
@@ -51,13 +52,15 @@ class WidgetsApp extends StatefulWidget {
     this.checkerboardRasterCacheImages: false,
     this.checkerboardOffscreenLayers: false,
     this.showSemanticsDebugger: false,
+    this.showWidgetInspector: false,
```
also happy to change all the names to 
WidgetDebugger from WidgetInspector.

jacob314 commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,444 @@
+
+  void _update() {
+    SchedulerBinding.instance.addPostFrameCallback((Duration timeStamp) {
+      // We want the update to take effect next frame, so to make that
```
This logic is from SemanticsDebugger. Not sure changes taking effect next frame makes any sense here. Let me know if there is a sensible way to trigger repainting at the end of every frame as that is really all I want.

Hixie replied:
what do you mean by the "end of every frame"?

jacob314 replied:
I mean I just want to repaint every frame where anything has actually changed.

jacob314 replied:
As per offline discussion switched to using layers instead which is exactly what I needed to get callbacks at the correct time.

jacob314 commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,444 @@
+    final RenderIgnorePointer ignorePointer =
+        _ignorePointerKey.currentContext.findRenderObject();
+    final RenderBox userRender = ignorePointer.child;
+    // TODO
```
It could make sense to iterate on the children of the hit test result to see if there is a child that also hit test so that we penetrate through RenderIgnorePointer and similar. Expect it would only make sense to do this under certain conditions. Perhaps a long click or clicking on an already selected widget should trigger this behavior.

Hixie replied:
if you did this, you'd have to apply the paint transform for the child from the parent before doing the hit test of the child.

But if you're going to do that, it might make sense to just do that uniformly, and not rely on the hit-testing logic at all. After all, you never know which render objects are going to be excluding arbitrary children from hit testing, even if they include others.

jacob314 replied:
Unfortunately, purely using bounding boxes results in unintuitive behavior even when correctly handling offstage children. 
I have a candidate solution that behaves well for sample applications by generating all matches excluding offstage elements but prioritizes matches that correspond to hit test matches. In addition, bounding boxes for all matches are shown and a match is prioritized if the cursor is on the edge the bounding box which allows for reasonable selection in the UI.

jacob314 commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,444 @@
+    setState(() {
+      if (result.path.isNotEmpty) {
+        _selected = result.path
+            .firstWhere(((HitTestEntry t) => t.target is RenderBox))
```
Not sure how important it is to handle non-render box hit targets here.

jacob314 replied:
Another thing to consider is it may be useful to outline other hit test results allowing users to cycle between them or at least understand what hit test matches are.

Hixie replied:
Well, if you're trying to debug a sliver, then not showing slivers will be a problem. If you're trying to debug a sector based layout, not showing sectors will be a problem. But I don't know how you'd represent those widgets when painting, either. Sectors in particular aren't rectangles, for example. Slivers are pretty complicated beasts too, with multiple relevant rectangular geometries (paintExtent, layoutExtent, hitTestExtent, etc).

jacob314 replied:
Fixed.

jacob314 commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,444 @@
+  final double wedgeSize = _kTooltipPadding * 2;
+  double wedgeX = math.max(tipOffset.dx, target.dx) + wedgeSize * 2;
+  wedgeX = math.min(wedgeX, tipOffset.dx + tooltipSize.width - wedgeSize * 2);
+  canvas.drawPath(
```
This is to render a chat bubble style tooltip like chrome devtools. Obviously if we decide this is a good UI to make it clear what object the tip is associated with we should polish this code as the border for the bubble isn't right.

jacob314 commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,444 @@
+
+  @override
+  bool shouldRepaint(_WidgetInspectorPainter oldDelegate) {
+    return isSelectionActive ||
```
this is overkill. should really be isSelectionActive || oldDelegate.isSelectionActive
I need to clean this code up but waiting as I suspect I'm missing some cleaner way to track RenderObject changes. Perhaps no need to optimize and we should just render every frame if there is a selection as this is just a debugging tool.

Hixie replied:
this function should compare the constructor arguments, which is all that the paint should be depending on. If the paint depends on anything outside of the constructor arguments, then even if this function always returns true, it won't repaint, because this function is only called when you recreate the delegate.

Hixie replied:
So in particular, for example, the paint method right now depends on "selected.getTransformTo(null)". But you can't do that, because that will change without notifying you.

jacob314 replied:
removed. Added corrected repaint logic to the layer code.

jacob314 commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,444 @@
+// FOR DISCUSSION: this is clearly not quite right: should I even even bother
+// listening for changes? This code is copied from semantics_debugger where
+// it made sense.
+class _WidgetClient extends ChangeNotifier {
```
Is there something similar I should be doing to listen for Widget changes?

Hixie replied:
i don't understand what this is trying to do in your case. Why do you care about the semantics tree?

jacob314 replied:
I don't. This code is pretty bogus. It only made sense for the semantic debugger. Is thee way I can get similar change notifications for Widget changes?

jacob314 replied:
What are your thoughts on the best way to get the notifications I need for this case?

Hixie replied:
I'm not sure I have context to know what you are asking.

jacob314 replied:
Never mind. Switching to managing my own Layer got what I was looking for.

jacob314 commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,444 @@
+  final int generation;
+  final RenderBox selected;
+
+  bool get isSelectionActive => selected != null && selected.attached;
```
if selected is not attached we should make a reasonable effort to see if there is an attached RenderBox that matches... e.g. same key or perhaps same class and path don't the render tree. Would be nice if you could scroll an element off screen and then back on and have the selection redisplay. Similarly, would be nice to continue showing the selection after as wide a range of hot reload edits as possible.

Hixie commented on packages/flutter/lib/src/widgets/app.dart:
```diff
@@ -252,6 +264,11 @@ class _WidgetsAppState extends State<WidgetsApp> implements WidgetsBindingObserv
         child: result
       );
     }
+    if (widget.showWidgetInspector || WidgetsApp.showWidgetInspectorOverride) {
```
We should probably consider limiting the semantics debugger and widget inspector to debug mode, so that they can get tree-shaken out.

jacob314 replied:
Done.

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,444 @@
+/// A widget that enables inspecting the child widget's structure.
+///
+/// This widget is useful for understand how an app is structured enabling
+/// interactive
```
nit: docs should have full grammar and punctuation, so there's a missing period here.

jacob314 replied:
Fixed by removing incomplete comment.

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,444 @@
+  /// Creates a widget that enables inspection for the child.
+  ///
+  /// The [child] argument must not be null.
+  const WidgetInspector({Key key, this.child}) : super(key: key);
```
nit: spaces around named arguments

jacob314 replied:
Done

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,444 @@
+  ///
+  /// In select mode pointer interactions trigger widget selection instead of
+  /// normal interactions.
+  /// Not in select mode, the selected widget is highlighted but the application
```
Paragraph breaks require a blank line. If you meant this to be three paragraphs, you need a new newline here. If you meant it to be two, you should probably rewrap the second paragraph to have consistent line lengths.

jacob314 replied:
done

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,444 @@
+  /// normal interactions.
+  /// Not in select mode, the selected widget is highlighted but the application
+  /// can be interacted with normally.
+  bool isSelectMode = false;
```
this should probably be private

jacob314 replied:
done

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,444 @@
+  /// can be interacted with normally.
+  bool isSelectMode = false;
+
+  /// Key tracking where the root in the widget tree
```
nit: missing period

Hixie replied:
actually, missing the rest of the sentence too :-)

jacob314 replied:
Removed sentence as member is private.

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,444 @@
+    // binding. Instead, we should find a way to get to the PipelineOwner from
+    // the BuildContext.
+    _client = new _WidgetClient(WidgetsBinding.instance.pipelineOwner)
+      ..addListener(_update);
```
i don't understand why you need the pipeline owner at all.

jacob314 replied:
I do not. Removed.

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,444 @@
+                  onPointerCancel: _handlePointerCancel,
+                  behavior: HitTestBehavior.opaque,
+                  child: isSelectMode
+                      ? new IgnorePointer(
```
Avoid adding and removing widgets like this, it causes the entire subtree to rebuild when the value is toggled, which is very expensive (and can be destructive to state if you didn't carefully wrap the child in a globalkey).

IgnorePointer has a flag you can just toggle (passing isSelectMode).

jacob314 replied:
done.

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,444 @@
+            left: _kInspectButtonMargin,
+            bottom: _kInspectButtonMargin,
+            child: isSelectMode
+                ? new Container()
```
just exclude this entire subtree if you don't want it (by not including it in the stack child list).

jacob314 replied:
Done

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,444 @@
+  Widget build(BuildContext context) {
+    return new CustomPaint(
+        foregroundPainter: new _WidgetInspectorPainter(
+            _pipelineOwner, _client.generation, _selected),
```
nit: (here and other places) when you have a newline after the `(`, have a newline before the `)`.

jacob314 replied:
Done. Audited all indentation in the file.

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,444 @@
+  void _handlePointerUp(PointerUpEvent event) {
+    _lastPointerLocation = event.position;
+
+    // If the pointer up position is on the edge of the window assume that it
```
mouse -> touch

jacob314 replied:
done

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,444 @@
+    // dragged off the edge of the display we do not want to select anything.
+    final Rect bounds =
+        (Offset.zero & (ui.window.physicalSize / ui.window.devicePixelRatio))
+            .deflate(_kOffScreenMargin);
```
does this mean there's no way to inspect something on the margin?

jacob314 replied:
Updated comment to explain that case.
    // A user can still select a widget that is only at the exact screen
    // margin by taping.

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,444 @@
+                  onPointerDown: _handlePointerDown,
+                  onPointerMove: _handlePointerMove,
+                  onPointerUp: _handlePointerUp,
+                  onPointerCancel: _handlePointerCancel,
```
having both a listener and a gesture detector is a bit weird, but ok

Hixie replied:
having both a listener and a gesture detector, the way you have it set up, is going to cause weird behaviour, because the gesture detector is going to try to compete in the arena to do gesture disambiguation, but the listener logic is ignoring the arena and just happily handling every pointer.

You probably want a MultiTap or MultiDrag gesture handler.

jacob314 replied:
Done

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,444 @@
+  }
+
+  void _handlePointerDown(PointerDownEvent event) {
+    _lastPointerLocation = event.position;
```
this logic doesn't handle multitouch. if you're going to use a Listener and directly handle pointer events, you need to track each pointer individually.

jacob314 replied:
Switched to just using a gesture recognizer. I don't know how interesting the multitouch use case is for an inspector. Potentially we could attempt to only find render objects within the bounds of all active touch locations but I suspect that would be more confusing than useful.

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,444 @@
+/// pointer is moving off the device.
+const double _kOffScreenMargin = 1.0;
+
+// TODO(jacobr): merge with similar code in tooltip.dart.
```
we should definitely do that before merging this in

jacob314 replied:
Merged.

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,444 @@
+  bool _tooltipBelow;
+
+  bool get tooltipBelow => _tooltipBelow;
+  Offset get childPosition => _childPosition;
```
if you want to use a class in this way, i would use a factory constructor so that the class can be const and you don't need any mutable fields

jacob314 replied:
done

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,444 @@
+  final String message = '$render\n\n'
+      '${render.debugCreator}';
+  canvas.save();
+  final TextPainter textPainter = new TextPainter()
```
TextPainters are expensive (specifically, the call to layout is expensive) so you want to only lay it out once per set of changes. creating a text painter in a paint call is to be avoided where possible.

jacob314 replied:
Now that I'm caching the output to a layer and only re-rendering on state change, does attempting to cache the TextPainter really matter? Any time the message or the width of the selected render object changes the painter would need to be updated anyway.

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,444 @@
+          new Offset(wedgeX - wedgeSize, wedgeY),
+          new Offset(wedgeX + wedgeSize, wedgeY),
+          new Offset(
+              wedgeX, wedgeY + (calc.tooltipBelow ? -wedgeSize : wedgeSize)),
```
another example of the `)` issue

jacob314 replied:
Did a pass through the file cleaning up non-flutter indentation.

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,444 @@
+  textPainter.paint(
+      canvas, tipOffset + const Offset(_kTooltipPadding, _kTooltipPadding));
+  canvas.restore();
+}
```
another example of the `)` issue

jacob314 replied:
done

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,444 @@
+      transform.transform3(new Vector3(0.0, 0.0, 0.0)),
+      transform.transform3(new Vector3(selected.size.width, 0.0, 0.0)),
+      transform.transform3(
+          new Vector3(selected.size.width, selected.size.height, 0.0)),
```
i don't think the newline helps readability here, it just makes the code less uniform

jacob314 replied:
Removed this code as I located the existing helper handling this case.

Hixie commented on packages/flutter_tools/lib/src/resident_runner.dart:
```diff
@@ -716,6 +732,7 @@ abstract class ResidentRunner {
       printStatus('You can dump the widget hierarchy of the app (debugDumpApp) by pressing "w".');
       printStatus('To dump the rendering tree of the app (debugDumpRenderTree), press "t".');
       printStatus('For layers (debugDumpLayerTree), use "L"; accessibility (debugDumpSemantics), "S".');
+      printStatus('To toggle the widget inspector (widgetInspector), press "i".');
```
The parenthetical here is supposed to be the thing you can look up in the docs to learn about it, and also the thing that you can toggle from code to do the same thing. So in this case, it should probably be `WidgetsApp.showWidgetInspectorOverride`.

jacob314 replied:
done.

Hixie commented:
The general approach seems fine. One major problem is that there's no way to know when the item you are watching is moving, so if you inspect something that's animating, your border won't move with it. To fix that you'll need something like https://github.com/flutter/flutter/issues/3861 (though I don't really know how you'd handle it even with that exactly). You could work around it by somehow requesting that you paint every frame that paints, but the system basically spends all its effort making that impossible because it'd be a performance disaster, so there's no good API to hook into to do that right now either.

Hixie commented:
It sounds like what you want to do is actually hook into the pipeline at a pretty low level. If you're ok with requiring that people add a line of code to their main.dart and import a separate package, you could provide a new binding that overrides the FlutterWidgetsBinding to provide the relevant hooks, in much the same way that the test framework does that. (In your case, you'd probably need to create a new BuildOwner so that you could override the handling of dirty widgets so that any time you have a dirty widget, you also mark the inspector as potentially needing an update, as well as creating a new PipelineOwner so you could catch any changes that result in a paint update, in case any of the render objects are repainting independent of any widget updates.)

The alternative would be to hook onDrawFrame to just always mark the inspector dirty, passing through every call to the underlying binding. That would be really sketchy though. For example, it would totally break down if you had multiple inspector widgets at once. It would also be less efficient, though honestly even the binding override won't be great in that respect.

Hixie commented:
@jacob314 Should we close this PR for now?

Hixie commented:
@jacob314 I'm going to close this PR since my understanding is that you're not awaiting review and do not plan to land this exact PR, and so we should take it off the review queue. Don't hesitate to reopen it or open a new one when you're ready for another round of review!

Hixie commented:
@jacob314 You reopened the PR, does that mean you want review again?

jacob314 commented:
I'm going through the comments again now. I need to address a few more of them before it is ready to be reviewed again.

jacob314 commented:
Using a layer appears to work great for determining when to redraw.
Pushed a new version which should address all comments.

Ready for a full review.

Hixie commented on packages/flutter/lib/src/material/tooltip.dart:
```diff
@@ -196,8 +196,9 @@ class _TooltipState extends State<Tooltip> with SingleTickerProviderStateMixin {
 
-class _TooltipPositionDelegate extends SingleChildLayoutDelegate {
-  _TooltipPositionDelegate({
+/// A delegate that positions a tooltip relative to a target.
```
if you want to make this public you're going to have to fill in all the documentation on this class :-)

Hixie replied:
("A delegate that positions a tooltip relative to a target" doesn't pass our style guide test of "the doc couldn't be written by someone who doesn't actually know anything about the class except its name", and the other members, like the constructor and fields, are lacking docs.)

jacob314 replied:
Documented.

Hixie commented on packages/flutter/lib/src/widgets/app.dart:
```diff
@@ -171,6 +178,11 @@ class WidgetsApp extends StatefulWidget {
 
+  /// If true, forces the widget inspector to be visible.
+  ///
+  /// Used by `showWidgetInspector` debugging extension.
```
grammar (missing "the" after "by"; please also fix the other cases in this file)

jacob314 replied:
done. fixed all the existing cases where "the" was omitted to bring in compliance with Hixie english. :)

Hixie commented on packages/flutter/lib/src/widgets/app.dart:
```diff
@@ -301,19 +313,25 @@ class _WidgetsAppState extends State<WidgetsApp> implements WidgetsBindingObserv
+        result = new WidgetInspector (
+          child: result,
+        );
+      }
```
if they're in the assert, that means they're only available in debug mode, in which case their names should start with `debug` and their docs should make this clear.

showSemanticsDebugger in particular is intended to work in profile builds too, so that one should stay outside the assert.

jacob314 replied:
done.

Hixie commented on packages/flutter/lib/src/widgets/app.dart:
```diff
@@ -171,6 +178,11 @@ class WidgetsApp extends StatefulWidget {
   /// Used by `showPerformanceOverlay` observatory extension.
   static bool showPerformanceOverlayOverride = false;
 
+  /// If true, forces the widget inspector to be visible.
```
This should say what the widget inspector is, how to use it, etc.

jacob314 replied:
done

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,526 @@
+import 'package:flutter/material.dart';
+import 'package:flutter/painting.dart';
+import 'package:flutter/rendering.dart';
+import 'package:vector_math/vector_math_64.dart';
```
you shouldn't need to import vector_math

jacob314 replied:
done.

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,526 @@
+import 'framework.dart';
+import 'gesture_detector.dart';
+
+/// A widget that enables inspecting the child widget's structure.
```
paragraphs of more details please

jacob314 replied:
done

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,526 @@
+    // pointer is being dragged off the edge of the display not a regular touch
+    // on the edge of the display. If the pointer is being dragged off the edge
+    // of the display we do not want to select anything. A user can still select
+    // a widget that is only at the exact screen margin by taping.
```
taping->tapping, i assume

jacob314 replied:
donne

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,526 @@
+      onPanEnd: _handlePanEnd,
+      onPanUpdate: _handlePanUpdate,
+      behavior: HitTestBehavior.opaque,
+      excludeFromSemantics: true,
```
How should we make this tool accessible?

jacob314 replied:
If we wanted to make it accessible, we should add inspector functionality to the semantics debugger itself as the semantics debugger actually makes it in some ways easier to select widgets as all objects have unique bounding boxes, etc.

Hixie replied:
Ironically, the semantics debugger isn't accessible either.

jacob314 replied:
Are there really any existing widget inspecting tools that are accessible? Seems like a very difficult problem to solve.

Hixie replied:
I don't know. Solving difficult problems is what we do, though. :-)

jacob314 replied:
you aren't suggesting we actually do anything about accessibility for the inspector in this pull request, correct?

Hixie replied:
Correct. I don't even know what it would really be. Please do file a bug with the a11y labels and projects so we can track it, though.

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,526 @@
+  /// Index within the list of candidates that is currently selected.
+  int index = 0;
+
+  set candidates(List<RenderObject> value) {
```
setter should be immediately after the getter which should be before the field.

jacob314 replied:
done

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,526 @@
+  /// The arguments must not be null.
+  _RenderInspectorOverlay({ @required this.selection }) : assert(selection != null);
+
+  _InspectorSelection selection;
```
should setting this call markNeedsPaint?

jacob314 replied:
why would it need to? The layer seems to get painted every frame anyway.
Perhaps performResize should call markNeedsPaint. Is that what you mean?

Hixie replied:
What if the only thing that happens during a frame is setting this? Then nothing will get marked dirty, and the layer won't repaint.

jacob314 replied:
done

Hixie replied:
Did you upload the patch with this fix?

Hixie replied:
Oh I see. Sorry, I wasn't clear. The markNeedsPaint call should be on the render object, not the render object widget. The "selection" property should be a getter/setter pair and the setter should call markNeedsPaint if the new value is different than the old. See other render objects for examples of the pattern.

jacob314 replied:
done.

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,526 @@
+
+  @override
+  bool operator ==(dynamic other) {
+    if (other is! _TransformedRect)
```
test runtimeType instead of using `is`.

jacob314 replied:
done

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,526 @@
+
+  @override
+  bool operator ==(dynamic other) {
+    if (other is! _InspectorOverlayRenderState)
```
ditto

jacob314 replied:
done

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,526 @@
+/// certain performance statistics within it.
+///
+/// Performance overlay layers are always leaves in the layer tree.
+class InspectorOverlayLayer extends Layer {
```
this can be private, right?

jacob314 replied:
yep. done

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,526 @@
+    final double verticalOffset = (targetRect.height) / 2 + offsetFromWidget;
+
+    _paintDescription(
+        canvas, state.tooltip, target, verticalOffset, size, targetRect);
```
remove the newline here, it's ok to have this on one line

jacob314 replied:
done

Hixie commented on packages/flutter/test/foundation/service_extensions_test.dart:
```diff
@@ -459,7 +459,7 @@ void main() {
     // If you add a service extension... TEST IT! :-)
     // ...then increment this number.
-    expect(binding.extensions.length, 15);
+    expect(binding.extensions.length, 16);
```
you also need to actually add a test! see the comment!

jacob314 replied:
done

Hixie commented on packages/flutter_tools/lib/src/vmservice.dart:
```diff
@@ -1090,6 +1090,19 @@ class Isolate extends ServiceObjectOwner {
   Future<Map<String, dynamic>> flutterTogglePerformanceOverlayOverride() => _flutterToggle('showPerformanceOverlay');
 
+  Future<Map<String, dynamic>> flutterToggleWidgetInspector() async {
+    Map<String, dynamic> state = await invokeFlutterExtensionRpcRaw('ext.flutter.widgetInspector');
```
why not just call _flutterToggle?

jacob314 replied:
no reason other then when this code was written. done.

Hixie commented:
This seems pretty solid. Only a few more tweaks. :-)

abarth commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,526 @@
+    _hitTestHelper(result, edgeHits, position, root, root.getTransformTo(null));
+    final Set<RenderObject> hits = new LinkedHashSet<RenderObject>();
+    hits.addAll(edgeHits);
+    result.forEach((RenderObject o, bool value) {
```
s/o/object/

(Style guide says to use complete words in variable names)

jacob314 replied:
done

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,526 @@
+void _paintDescription(Canvas canvas, String message, Offset target,
+    double verticalOffset, Size size, Rect targetRect) {
+  canvas.save();
+  final TextPainter textPainter = new TextPainter()
```
it seems unlikely to me that the text painter changes every frame (i mean, that would mean the text was unreadable, changing nonstop). This should be cached. Painting text is ridiculously expensive.

jacob314 replied:
Done. The text painter is now cached as it really only changes if the device is rotated or the selected RenderObject changes.

Hixie commented:
Given how expensive this approach is, we should make sure that this widget cannot run outside of debug mode. Can you add some code that will throw an exception in profile and release builds, saying that this widget isn't available outside debug mode? Also, please add a detailed comment to the Layer subclass explaining that this approach is horrifically for performance and is only used here because this is limited to debug mode and that nobody should think of using this in production code. :-)

abarth commented:
I'm not sure whether it is feasible, but here's a design that might be more efficient:

1. Move the widget tree that is under inspection to an off-screen pipeline owner.
2. When the off-screen pipeline owner requests a visual update, request a visual update in the on-screen pipeline owner.
3. In a persistent frame callback, flush the off-screen pipeline owner.
4. Read the relevant information from the off-screen pipeline owner and generate writes into a normal on-screen render tree.
5. Let the on-screen pipeline owner flush the on-screen render tree as usual.

This approach will let the inspector use all the machinery of the whole system to work efficiently instead of attempting to do its own layout and paint phase during the addToScene walk of the composite phase.

Essentially, using a separate pipeline owner separates the work from the inspector's pipeline and the tree under inspection so that they can run in series instead of in parallel.

If it is too disruptive to move the tree that is under inspection off-screen, another approach would be for the inspector to build a render tree in an off-screen pipeline owner and then flush the inspector's pipeline during the composite phase.  This approach is less elegant in the sense that we'd still be running a mini version of the whole system inside the composite phase of the real system, but at least you'd have all the machinery of the whole system available to you so you wouldn't need to do grossly inefficient things like re-creating TextPainter objects for every animation frame.

jacob314 committed:
Code review comment fixes

jacob314 commented:
Addressed all comments.
@abarth interesting ideas for how to improve the performance. I'd like to investigate the first option once the initial CL lands. Having a custom PipelineOwner could be interesting for other reasons as wel.
Keep in mind the current design at feels like it has fine performance for the applications I've tested with likely because the UI I need to render in the overlay is extremely simple. Note that the TextPainter is now cached almost all times when the selected RenderObject does not change. The current UI is as complex as the UI rendered within the screen area for Chrome Devtools so I expect there is a good change we do not need to render anything more sophisticated on the device screen to provide a good developer experience.

abarth commented:
Please don't take my comments as blocking in any way.  It's just an idea for an approach that might be more scalable (assuming it works at all).

> the UI I need to render in the overlay is extremely simple

These tools tend to get more sophisticated over time.  There's some value in choosing an approach with a solid foundation to be ready for the feature to grow over time.

Hixie commented:
@jacob314 Looks like you didn't push your recent changes (or at least, the first comment, which you say is done, doesn't seem to have been done).

jacob314 commented:
Fixed the typo in the git push command. Ready for review now.

jacob314 committed:
Break cyclic dependency.

Code review comment fixes.

jacob314 commented:
Removed the direct dependency on material by moving the tooltip code to /rendering and injecting a builder to create the button to re-enter inspect mode. If no builder is specified, the inspector always stays in select mode which is fine as toggling back and forth between select and view modes is an experiment which we could easily remove.

jacob314 commented:
Note that the hit testing code was also simplified to eliminate all dependence on RenderBox hit testing as RenderBox hit testing started giving undesirable results for sample applications due to an added gesture recognizer. The new heuristic is the render object with the smaller bounding rectangle wins if multiple hit test match occurs and the match is not on the edge of one of the rectangles.

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,553 @@
+import 'framework.dart';
+import 'gesture_detector.dart';
+
+typedef Widget InspectorSelectButtonBuilder(VoidCallback onPressed);
```
a Builder should take a BuildContext

jacob314 replied:
done.

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,553 @@
+  /// The widget that is being inspected.
+  final Widget child;
+
+  final InspectorSelectButtonBuilder _selectButtonBuilder;
```
why is this private?

jacob314 replied:
Because I don't want anyone outside of this library to depend on the fact that a InspectorWidget has a _selectButtonBuilder member.

Hixie replied:
why not? They create it, and pass it the value, so...

Normally widgets have all their constructor arguments just be public fields.

jacob314 replied:
Made public. I don't feel strongly about hiding this implementation detail.

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,553 @@
+    }
+
+    final Size tooltipSize = _textPainter.size + const Offset(_kTooltipPadding * 2, _kTooltipPadding * 2);
+    final TooltipPositionDelegate tooltipPositionDelegate = new TooltipPositionDelegate(
```
Creating a custom layout delegate when you really just want to call a function doesn't really make sense. Maybe we should instead leave the custom layout delegate where it was, and factor out the math into a reusable function in the painting layer?

I hadn't realised you weren't actually using a custom layout widget here.

Adam's suggestion of actually using a real pipeline seems really solid. It would let you remove all these hacks and just implement a real widget UI with all of the layout primitives and so on.

jacob314 replied:
I agree that switching to a real pipeline would be better and I'd like to accomplish that with a follow on CL that does the significant refactoring required to structure the design that way.
For now, refactored _TooltipPositionDelegate to use a top level method that computes the tooltip layout.

Hixie commented:
Are there any tests for the inspector itself?

jacob314 committed:
Add tests.

jacob314 commented:
Added tests to the inspector. Made  _WidgetInspectorState and _InspectorSelection public for easier testing.

jacob314 commented:
Ready for next round of review. Add tests for the inspector itself, addressed all other outstanding comments.

Hixie commented on packages/flutter/lib/src/painting/utils.dart:
```diff
@@ -0,0 +1,55 @@
+///
+/// Used by [Tooltip] to position a tooltip relative to its parent.
+///
+/// The arguments cannot be null.
```
s/cannot/must not/

jacob314 replied:
Fixed here and sent out https://github.com/flutter/flutter/pull/11683
so that the next person documenting non-nullable arguments isn't tripped up by
existing examples of "cannot be null" that should have been "must not be null".

Hixie commented on packages/flutter/lib/src/widgets/app.dart:
```diff
@@ -58,7 +59,9 @@ class WidgetsApp extends StatefulWidget {
-    this.debugShowCheckedModeBanner: true
+    this.debugShowWidgetInspector: false,
+    this.debugShowCheckedModeBanner: true,
+    InspectorSelectButtonBuilder inspectorSelectButtonBuilder,
```
why isn't this just `this.inspectorSelectButtonBuilder`?

jacob314 replied:
made the member public as there isn't a great reason for it to be private.

Hixie commented on packages/flutter/lib/src/widgets/app.dart:
```diff
@@ -168,17 +180,29 @@ class WidgetsApp extends StatefulWidget {
   /// with "s".
   static bool debugAllowBannerOverride = true;
 
+  final InspectorSelectButtonBuilder _inspectorSelectButtonBuilder;
```
This should be public (and in the same order and position as other constructor arguments).

jacob314 replied:
done. Also added documentation.

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,560 @@
+/// selecting the widget even if another widget that also overlaps that
+/// location would otherwise have priority.
+///
+/// In the view mode, the previous;y selected widget is outlined however
```
typo s/;/l/

Hixie replied:
also, commas around "however"

jacob314 replied:
done,

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,560 @@
+/// In the view mode, the previous;y selected widget is outlined however
+/// touching the device has the same effect it would have if the inspector
+/// wasn't present. This allows interacting with the application and viewing how
+/// the selected widget changes position. Clicking on the select icon in th
```
s/th/the/

jacob314 replied:
done

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,560 @@
+import 'framework.dart';
+import 'gesture_detector.dart';
+
+typedef Widget InspectorSelectButtonBuilder(BuildContext context, VoidCallback onPressed);
```
nit: needs docs

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,560 @@
+  /// The widget that is being inspected.
+  final Widget child;
+
+  /// A builder that is called to create the select button
```
nit: period at end of sentence

also, document the contract (in particular that the resulting widget tree should be hooked to the given onPressed)

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,560 @@
+  WidgetInspectorState createState() => new WidgetInspectorState();
+}
+
+class WidgetInspectorState extends State<WidgetInspector>
```
This should probably be private.

Alternatively, it and all its members have to be documented, but that doesn't seem useful to the end user.

You can work around it being private in the test by assigning it to a variable of type "dynamic" and then blindly accessing its members. (Just make sure to leave a comment explaining that it's ok that you're doing such a crazy thing because it's only a test!)

jacob314 replied:
done

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,560 @@
+    final Offset localPosition = MatrixUtils.transformPoint(inverse, position);
+
+    final List<DiagnosticsNode> children = object.debugDescribeChildren();
+    for (int i = children.length - 1; i >= 0; i--) {
```
nit: `--i` or, even better, `i -= 1`

jacob314 replied:
done. Switched to i -= 1 for consistency with other flutter code.

Hixie commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -0,0 +1,560 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
```
you need to export this file from ../widgets.dart

jacob314 replied:
Donedone

Hixie commented:
Modulo the minor comments:
![LGTM](http://junkyard.damowmow.com/579)

jacob314 committed:
Final round of code review comment fixes.
